### PR TITLE
support auto delete WorkloadRebalancer when time up

### PR DESCRIPTION
**Pull request by chaosi-zju**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Support auto delete WorkloadRebalancer when time up:

referring to [Automatic Cleanup for Finished Jobs](https://kubernetes.io/docs/concepts/workloads/controllers/ttlafterfinished/).

Introduces field `ttlSecondsAfterFinished` which limits the lifetime of a WorkloadRebalancer that has finished execution 
(finished execution means each target workload is finished with result of `Successful` or `Failed`).

* If this field is set, `ttlSecondsAfterFinished` after the WorkloadRebalancer finishes, it is eligible to be automatically deleted.
* If this field is unset, the WorkloadRebalancer won't be automatically deleted.
* If this field is set to zero, the WorkloadRebalancer becomes eligible to be deleted immediately after it finishes.

Considering several corner cases:

* case 1: if a new target workload was added into `WorkloadRebalancer` before `ttlSecondsAfterFinished` expired, 
  which means the finish time of the `WorkloadRebalancer` is refreshed, so the `delete` action is deferred since expire time is refreshed too.
* case 2: if `ttlSecondsAfterFinished` is modified before `ttlSecondsAfterFinished` expired, 
  `delete` action should be performed according to latest `ttlSecondsAfterFinished`.
* case 3: when we have got and checked latest `WorkloadRebalancer` object and try to delete it, 
  if a modification to `WorkloadRebalancer` occurred just right between the two time point, the previous `delete` action should be Interrupted.

Several key implementation:
* A `WorkloadRebalancer` is judged as finished should meet two requirements:
  *  all expected workloads are finished with result of `Successful` or `Failed`.
  *  introduce a new field named `ObservedGeneration` to `Status` of WorkloadRebalancer, and it should be equal to `.metadata.Generation`,
     to prevent that the WorkloadRebalancer is updated but controller hasn't in time refreshed its `Status`.
* When `WorkloadRebalancer` is `Created` or `Updated`, add it to the workqueue and calculate its expiring time, and 
  call `workqueue.AddAfter()` function to re-enqueue it once more if it hasn't expired.
* Before deleting the `WorkloadRebalancer`, do a final sanity check. Use the latest `WorkloadRebalancer` directly 
  fetched from api server to see if the TTL truly expires, rather than object from lister cache.
* When deleting the `WorkloadRebalancer`, it is needed to confirm that the `resourceVersion` of the deleted object is as expected,
  to prevent from above corner case 3.

**Which issue(s) this PR fixes**:

Fixes part of #4840

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
support auto delete WorkloadRebalancer when time up
```

### CODE REVIEW

**Comment by chaosi-zju** [comment]
/hold

for #4875 #4860

**Comment by chaosi-zju** [comment]
CC @RainbowMango can this begin to review?

File: pkg/apis/apps/v1alpha1/workloadrebalancer_types.go
Comment (RainbowMango) [inline pkg/apis/apps/v1alpha1/workloadrebalancer_types.go:67]: Just a question, why not use `seconds` but `minutes` here?

File: pkg/controllers/workloadrebalancer/ttlafterfinished_worker.go
Comment (RainbowMango) [inline pkg/controllers/workloadrebalancer/ttlafterfinished_worker.go:108]: Can you show me where to set the `LastUpdateTime`?

File: pkg/controllers/workloadrebalancer/workloadrebalancer_controller.go
Comment (RainbowMango) [inline pkg/controllers/workloadrebalancer/workloadrebalancer_controller.go:54]: Smells not right, can't we check and remove the object in the reconcile loop?

File: pkg/apis/apps/v1alpha1/workloadrebalancer_types.go
Comment (RainbowMango) [inline pkg/apis/apps/v1alpha1/workloadrebalancer_types.go:105]: Do we need to update this `LastUpdateTime` during each reconcile loop?

File: pkg/controllers/workloadrebalancer/ttlafterfinished_worker.go
Comment (chaosi-zju) [reply pkg/controllers/workloadrebalancer/ttlafterfinished_worker.go:108]: Sorry, it is my fault when my last `rebase` process, I encountered many conflict.

When I resolving the conflict, I missed the key line to update `LastUpdateTime`.

https://github.com/karmada-io/karmada/compare/7b02e4e882620de6b4f41fff6d8262c43755b669..32819f2972c056ee53a2a51ebbf0d1f6bf7e7cc9

image

I will do a quick fix.

File: pkg/apis/apps/v1alpha1/workloadrebalancer_types.go
Comment (chaosi-zju) [reply pkg/apis/apps/v1alpha1/workloadrebalancer_types.go:105]: when any field in `WorkloadRebalancerStatus` other than `LastUpdateTime` itself changed, we should update this field.

File: pkg/apis/apps/v1alpha1/workloadrebalancer_types.go
Comment (chaosi-zju) [reply pkg/apis/apps/v1alpha1/workloadrebalancer_types.go:67]: In my expected usage scenarios of `WorkloadRebalancer`, users do not need to be accurate to seconds, only to minutes.

I think `rebalance` is a key operation. In actual usge, it may take more than ten minutes to observe whether the status of the instance meets expectations after the operation. If there is an exception, it may take another ten minutes to use this resource object to locate the problem. So the deletion time must be in minutes.

Comment (chaosi-zju) [reply pkg/apis/apps/v1alpha1/workloadrebalancer_types.go:67]: besides, since the usage scenario is `minites` level, `minutes` may be more readable than `seconds`.

File: pkg/controllers/workloadrebalancer/workloadrebalancer_controller.go
Comment (chaosi-zju) [reply pkg/controllers/workloadrebalancer/workloadrebalancer_controller.go:54]: I didn't understand, can you give an example?

File: pkg/apis/apps/v1alpha1/workloadrebalancer_types.go
Comment (chaosi-zju) [reply pkg/apis/apps/v1alpha1/workloadrebalancer_types.go]: Updated to `finishTime`.

File: pkg/controllers/workloadrebalancer/workloadrebalancer_controller.go
Comment (chaosi-zju) [reply pkg/controllers/workloadrebalancer/workloadrebalancer_controller.go:54]: ready to continue to review~ @RainbowMango

File: pkg/apis/apps/v1alpha1/workloadrebalancer_types.go
Comment (RainbowMango) [inline pkg/apis/apps/v1alpha1/workloadrebalancer_types.go:61]:
```suggestion
	// TTLSecondsAfterFinished limits the lifetime of a WorkloadRebalancer that has finished execution (means each
```

File: pkg/apis/apps/v1alpha1/workloadrebalancer_types.go
Comment (RainbowMango) [inline pkg/apis/apps/v1alpha1/workloadrebalancer_types.go:100]:
```suggestion
	// +optional
```

File: pkg/apis/apps/v1alpha1/workloadrebalancer_types.go
Comment (RainbowMango) [inline pkg/apis/apps/v1alpha1/workloadrebalancer_types.go:104]:
```suggestion
	// +optional
```

**Review by RainbowMango — COMMENTED** [review]
Don't have enough time to finish the review, just covered the API part this time.
Will get back on this next week.

File: pkg/controllers/workloadrebalancer/workloadrebalancer_controller.go
Comment (RainbowMango) [inline pkg/controllers/workloadrebalancer/workloadrebalancer_controller.go:52]: Here looks weird to have two control plane clients, one is used to fetch resources from the cache, and the other bypasses the cache.
Do we need to do that in such a way?

By the way, the comments are not precise enough, as the client(comes from controller-manager framework) not only operates resources against the cache.

File: pkg/controllers/workloadrebalancer/ttlafterfinished_worker.go
Comment (RainbowMango) [inline pkg/controllers/workloadrebalancer/ttlafterfinished_worker.go:65]: I can understand that you don't 100% trust the data from the cache, so you fetch the object directly from the api-server.

But the object can be any time, consider the situation that the updation happened just after you fetch the object, and before deleting, the updation will be dropped, right?

File: pkg/controllers/workloadrebalancer/workloadrebalancer_controller.go
Comment (RainbowMango) [inline pkg/controllers/workloadrebalancer/workloadrebalancer_controller.go:79]: I am still not sure if we should introduce another queue to check the TTL yet, just feel a little bit heavy. I will look at the Job controller from Kubernetes, and then get back on this.

File: pkg/controllers/workloadrebalancer/ttlafterfinished_worker.go
Comment (chaosi-zju) [reply pkg/controllers/workloadrebalancer/ttlafterfinished_worker.go:65]: > consider the situation that the updation happened just after you fetch the object, and before deleting, the updation will be dropped, right?

**No, it will not be droped, the delete action will wait for updation reconciled.**

 a corner, supposing the updation is modify `ttlSecondsAfterFinished` from 10s to 100s, the real delete action will defer to 100s after the latest updation.

by the way, this double check is also refer to job controller https://github.com/kubernetes/kubernetes/blob/a31030543c47aac36cf323b885cfb6d8b0a2435f/pkg/controller/ttlafterfinished/ttlafterfinished_controller.go#L206-L241

File: pkg/controllers/workloadrebalancer/workloadrebalancer_controller.go
Comment (chaosi-zju) [reply pkg/controllers/workloadrebalancer/workloadrebalancer_controller.go:52]: >  the comments are not precise enough, as the client(comes from controller-manager framework) not only operates resources against the cache.

got it, thanks.

> Here looks weird to have two control plane clients, ... Do we need to do that in such a way?

this is the same problem as next comment below.

Comment (RainbowMango) [reply pkg/controllers/workloadrebalancer/workloadrebalancer_controller.go:52]:
```suggestion
	Client             client.Client
	ClientWithOutCache versioned.Interface
```
How about using the raw client instead of the one from the controller-runtime framework?

File: cmd/controller-manager/app/controllermanager.go
Comment (RainbowMango) [inline cmd/controller-manager/app/controllermanager.go:715]:
```suggestion
		ClientWithOutCache: versioned.NewForConfigOrDie(ctx.Mgr.GetConfig()),
```

File: pkg/controllers/context/context.go
Comment (RainbowMango) [inline pkg/controllers/context/context.go:113]:
```suggestion
```
Then, we don't need to introduce the unusual client to the context.

**Comment by chaosi-zju** [comment]
CC @RainbowMango ready to continue

besides, I modified a little code in e2e test, please help me do a check @XiShanYongYe-Chang

File: pkg/controllers/workloadrebalancer/ttlafterfinished_worker.go
Comment (RainbowMango) [inline pkg/controllers/workloadrebalancer/ttlafterfinished_worker.go:75]: The delete option(`ResourceVersion`) reminds me that we might not need to fetch the object from the APIServer.
Since in case of the object is rotten, the delete option would prevent this object from deleting.

/hold
Please help to run a test for it to prove it is necessary.

File: cmd/controller-manager/app/controllermanager.go
Comment (RainbowMango) [inline cmd/controller-manager/app/controllermanager.go:714]: might be a lint issue.

File: pkg/controllers/workloadrebalancer/workloadrebalancer_controller.go
Comment (chaosi-zju) [inline pkg/controllers/workloadrebalancer/workloadrebalancer_controller.go:236]: Here is a bug, `rebalancer.Status` and `newStatus` will always not equal, which is not expected.

1. `rebalancer.Status.FinishTime` may be not nil, but `newStatus` will always be nil, we have not assigned a value to `newStatus.FinishTime`.
2. `rebalancer.Status` is a struct, while `newStatus` is a pointer.

This will lead to: even though `rebalancer.Status` unchanged, we will update `rebalancer.Status.FinishTime` when the rebalancer is reconciled.

File: pkg/controllers/workloadrebalancer/ttlafterfinished_worker.go
Comment (chaosi-zju) [reply pkg/controllers/workloadrebalancer/ttlafterfinished_worker.go]: done, updated

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve

**Comment by RainbowMango** [comment]
/hold
Please @XiShanYongYe-Chang take another look, particularly the e2e part.

**Comment by chaosi-zju** [comment]
@XiShanYongYe-Chang not only e2e, but also ut

**Comment by chaosi-zju** [comment]
@RainbowMango sorry, I did a updation after you send lgtm: https://github.com/karmada-io/karmada/compare/e405822231187e59ef62505a349aea75b8411ed2..f5076cdbeeb796d7df6426beb7ad98a1e790ceac

**Comment by XiShanYongYe-Chang** [comment]
Ok~
/assign

File: pkg/controllers/workloadrebalancer/workloadrebalancer_controller.go
Comment (XiShanYongYe-Chang) [inline pkg/controllers/workloadrebalancer/workloadrebalancer_controller.go:104]: Is this line of log really effective in locating the problem or is the log level too high? Logs about the processing failure of each workload are generated.

File: pkg/controllers/workloadrebalancer/workloadrebalancer_controller.go
Comment (XiShanYongYe-Chang) [inline pkg/controllers/workloadrebalancer/workloadrebalancer_controller.go:97]: Is step 3 part of step 2, or can the status update be put into method `doWorkloadRebalance`?

Comment (chaosi-zju) [reply pkg/controllers/workloadrebalancer/workloadrebalancer_controller.go:97]: I don't think step 3 is part of step 2, these 3 steps are parallel: read rebalancer -> work according to rebalancer -> write rebalancer

File: pkg/controllers/workloadrebalancer/workloadrebalancer_controller.go
Comment (XiShanYongYe-Chang) [inline pkg/controllers/workloadrebalancer/workloadrebalancer_controller.go:113]: This judgement `rebalancer.Status.FinishTime != nil` seems will always be true.

File: pkg/controllers/workloadrebalancer/workloadrebalancer_controller.go
Comment (XiShanYongYe-Chang) [inline pkg/controllers/workloadrebalancer/workloadrebalancer_controller.go:275]: Learned a new skill, thanks.

File: pkg/apis/apps/v1alpha1/workloadrebalancer_types.go
Comment (XiShanYongYe-Chang) [inline pkg/apis/apps/v1alpha1/workloadrebalancer_types.go:67]: Can we point out that the unit of time here is seconds?

File: pkg/controllers/workloadrebalancer/workloadrebalancer_controller.go
Comment (chaosi-zju) [reply pkg/controllers/workloadrebalancer/workloadrebalancer_controller.go:97]: done

File: test/e2e/framework/workloadrebalancer.go
Comment (XiShanYongYe-Chang) [inline test/e2e/framework/workloadrebalancer.go:93]: Use `return nil` maybe more readable.

**Review by XiShanYongYe-Chang — COMMENTED** [review]
Thanks~
Generally OK.

File: pkg/controllers/workloadrebalancer/workloadrebalancer_controller.go
Comment (chaosi-zju) [reply pkg/controllers/workloadrebalancer/workloadrebalancer_controller.go]: updated

**Comment by RainbowMango** [comment]
> [FAILED] in [DeferCleanup (Each)] - /home/runner/work/karmada/karmada/test/e2e/framework/workloadrebalancer.go:46 @ 05/25/24 10:13:54.839

It seems related, see [the logs here](https://github.com/karmada-io/karmada/actions/runs/9234550079/job/25408399107?pr=4894#step:6:3815).

```
[FAILED] Unexpected error:
      <*errors.StatusError | 0xc0005c8d20>: 
      workloadrebalancers.apps.karmada.io "rebalancer-lhnmp" not found
      {
          ErrStatus: {
              TypeMeta: {Kind: "", APIVersion: ""},
              ListMeta: {
                  SelfLink: "",
                  ResourceVersion: "",
                  Continue: "",
                  RemainingItemCount: nil,
              },
              Status: "Failure",
              Message: "workloadrebalancers.apps.karmada.io \"rebalancer-lhnmp\" not found",
              Reason: "NotFound",
              Details: {
                  Name: "rebalancer-lhnmp",
                  Group: "apps.karmada.io",
                  Kind: "workloadrebalancers",
                  UID: "",
                  Causes: nil,
                  RetryAfterSeconds: 0,
              },
              Code: 404,
          },
      }
  occurred
  In [DeferCleanup (Each)] at: /home/runner/work/karmada/karmada/test/e2e/framework/workloadrebalancer.go:46 @ 05/25/24 10:13:54.839
```

**Comment by chaosi-zju** [comment]
CI problem  fixed, is there any further comments?

**Comment by XiShanYongYe-Chang** [comment]
/lgtm
Thanks

**Pull request merged**
